test/Microsoft.ML.OnnxTransformTest/DnnImageFeaturizerTest.cs

```diff
@@ -0,0 +1,204 @@
+        public void OnnxStatic()
+        {
+            if (!RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+                return;
+
+            using (var env = new ConsoleEnvironment(null, false, 0, 1, null, null))
+            {
+                var imageHeight = 224;
+                var imageWidth = 224;
+                var dataFile = GetDataPath("images/images.tsv");
+                var imageFolder = Path.GetDirectoryName(dataFile);
+
+                var data = TextLoader.CreateReader(env, ctx => (
+                    imagePath: ctx.LoadText(0),
+                    name: ctx.LoadText(1)))
+                    .Read(dataFile);
+
+                var appDataBaseDir = Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData);
+                var execDir = Path.Combine(appDataBaseDir, "mlnet-resources");
+                var pipe = data.MakeNewEstimator()
+                    .Append(row => (
+                        row.name,
+                        data_0: row.imagePath.LoadAsImage(imageFolder).Resize(imageHeight, imageWidth).ExtractPixels(interleaveArgb: true)))
+                    .Append(row => (row.name, output_1: row.data_0.DnnImageFeaturizer((m, envi, input, estimator_output) => m.ResNet18(envi, input, estimator_output, execDir))));
+
+                TestEstimatorCore(pipe.AsDynamic, data.AsDynamic);
+
+                var result = pipe.Fit(data).Transform(data).AsDynamic;
+                result.Schema.TryGetColumnIndex("output_1", out int output);
+                using (var cursor = result.GetRowCursor(col => col == output))
+                {
+                    var buffer = default(VBuffer<float>);
+                    var getter = cursor.GetGetter<VBuffer<float>>(output);
+                    var numRows = 0;
+                    while (cursor.MoveNext())
+                    {
+                        getter(ref buffer);
+                        Assert.Equal(512, buffer.Length);
+                        numRows += 1;
+                    }
+                    Assert.Equal(3, numRows);
+                }
+            }
+        }
+
+        [ConditionalFact(typeof(Environment), nameof(Environment.Is64BitProcess))] // x86 fails with "An attempt was made to load a program with an incorrect format."
+        void TestOldSavingAndLoading()
+        {
+            if (!RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+                return;
+            
+
+            var samplevector = getSampleArrayData();
+
+            var dataView = ComponentCreation.CreateDataView(Env,
+                new TestData[] {
+                    new TestData()
+                    {
+                        data_0 = samplevector
+                    }
+                });
+
+            var inputNames = "data_0";
+            var outputNames = "output_1";
+            var appDataBaseDir = Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData);
```

Comment:
We shouldn't be using `LocalApplicationData` for these resources. Instead, since they are shipped as part of a NuGet package, the resource should be available from that NuGet package.

Reply:
This was done for cases where someone clones and builds the repo directly. Correct me if this is wrong, but if someone just clones the repo directly then they won't have the NuGet package. They might choose to generate it while building, or they might not. Regardless, the model needs to be downloaded somewhere if the repo is cloned directly, and LocalApplicationData seemed like a good choice. The reason why I also chose to use this in the tests is because the tests usually directly reference other projects, instead of downloading the corresponding NuGet package (as that package might not even exist on the server yet until a "release" happens). So I was under the impression that the tests did not have access to the NuGet packages but rather just the source code which they compile themselves. And that source code from the repo does not contain the models.

---
In reply to: [232768937](https://github.com/dotnet/machinelearning/pull/1447#discussion_r232768937) [](ancestors = 232768937)

Reply:
The way we do this for other native assets is to copy them from the `bin` directory to the test output directory during the build of the test.

So the workflow should look like this:

1. Building 'src' will download the asset to somewhere under the `bin\obj` directory.
2. The above location is where the NuGet package will package it from.
3. Building a `test` that depends on the assets, will copy the asset from under the `bin\obj` directory into the test's runtime folder. The test can just reference it like a customer's console app would be able to - get the file in the current directory of the "app".

In general, it is not a good idea to change machine state during a build. If someone clones the repo, then builds it. Deleting the repo (or `git clean`) should be good enough to "clean" all the assets from the repo.

Reply:
Reworked to be exactly as you recommended. Thanks for the detailed description!

---
In reply to: [232796081](https://github.com/dotnet/machinelearning/pull/1447#discussion_r232796081) [](ancestors = 232796081)